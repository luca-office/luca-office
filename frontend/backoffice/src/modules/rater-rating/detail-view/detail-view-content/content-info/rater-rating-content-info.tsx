import {css} from "@emotion/react"
import * as React from "react"
import {Heading, Text} from "shared/components"
import {HeadingLevel} from "shared/enums"
import {SurveyLight} from "shared/models"
import {
  Flex,
  FontWeight,
  spacingLarge,
  spacingMedium,
  spacingTiny,
  subHeaderHeight,
  textEllipsis,
  TextSize
} from "shared/styles"
import {useLucaTranslation} from "shared/translations"
import {Option} from "shared/utils"
import {getManualSurveyTimingHeadline, getSurveyTimingHeadline} from "../../../../../utils"
import {isManualSurvey} from "../../../../surveys/utils/common"

export interface RaterRatingContentInfoProps {
  readonly survey: SurveyLight
}

export const RaterRatingContentInfo: React.FC<RaterRatingContentInfoProps> = ({survey}) => {
  const {t} = useLucaTranslation()

  const isManualExecutionType = isManualSurvey(Option.of(survey.executionType))

  const surveyStatusLabel = isManualExecutionType
    ? getManualSurveyTimingHeadline(t, survey)
    : getSurveyTimingHeadline(t, survey.startsAt, survey.endsAt, survey.isCompleted)

  return (
    <div css={styles.info}>
      <div css={styles.infoContentWrapper}>
        <Text customStyles={styles.label} size={TextSize.Medium}>
          {t("rater_rating_details__title")}
        </Text>
        <div css={styles.infoTitle}>
          <Heading customStyles={styles.infoTitleLabel} level={HeadingLevel.h1}>
            {survey.title}
          </Heading>
          <Heading customStyles={textEllipsis} level={HeadingLevel.h1}>
            {surveyStatusLabel}
          </Heading>
        </div>
      </div>
      <div css={styles.infoContentWrapper}>
        <Text customStyles={styles.label} size={TextSize.Medium}>
          {t("rater_rating_details__description")}
        </Text>
        <Text customStyles={styles.description} size={TextSize.Medium}>
          {survey.description}
        </Text>
      </div>
    </div>
  )
}

const styles = {
  infoContentWrapper: css({
    display: "grid",
    gridTemplateRows: "minmax(min-content, max-content) 1fr",
    gridRowGap: spacingTiny
  }),
  label: css(textEllipsis, {
    fontWeight: FontWeight.Bold
  }),
  info: css({
    display: "grid",
    gridTemplateRows: "repeat(2, minmax(min-content, max-content))",
    gridRowGap: spacingLarge
  }),
  infoTitle: css(Flex.row, {
    justifyContent: "space-between"
  }),
  infoTitleLabel: css(textEllipsis, {
    marginRight: spacingMedium
  }),
  description: css({
    height: subHeaderHeight,
    overflow: "auto"
  })
}
