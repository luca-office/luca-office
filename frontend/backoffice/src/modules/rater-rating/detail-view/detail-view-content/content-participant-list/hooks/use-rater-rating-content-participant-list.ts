import {useApolloClient} from "@apollo/client"
import {isEqual, sumBy} from "lodash-es"
import React from "react"
import {useCodingCriteriaByItemsList, useScenarioCodingItemRatingsByRatingsList} from "shared/components"
import {getCodingItemsFromCodingDimensions} from "shared/components/rating/utils"
import {ProjectModuleProgressType, ProjectModuleType} from "shared/graphql/generated/globalTypes"
import {useProjectModules, useRatings, useSurveyInvitations} from "shared/graphql/hooks"
import {
  AutomatedCodingCriterion,
  CodingCriterion,
  CodingDimension,
  CodingItem,
  ProjectModule,
  Rating,
  ScenarioCodingItemRating,
  SurveyInvitationLight,
  SurveyLight
} from "shared/models"
import {every, find, flatten, sortByPosition} from "shared/utils"
import {
  isParticipantFullyRatedByScenario,
  isParticipantRatedByQuestionnaire as checkIsParticipantRatedByQuestionnaire
} from "../../../../../rating/utils"
import {FreetextQuestionRatingsForParticipants, useFreetextQuestionRatingsForParticipants} from "../../../../hooks"
import {RatingByParticipant, RatingProjectModule} from "../../../../models"
import {getCodingDimensions} from "../../../../utils"

export interface UseRaterRatingContentParticipantListHook {
  readonly dataLoading: boolean
  readonly participantCount: number
  readonly fullyRatedParticipantsCount: number
  readonly projectModulesCount: number
  readonly fullyRatedProjectModulesCount: number
  readonly surveyInvitations: SurveyInvitationLight[]
  readonly ratingProjectModules: RatingProjectModule[]
  readonly allCodingDimensions: CodingDimension[]
}

export const useRaterRatingContentParticipantList = (
  userAccountId: UUID,
  survey: SurveyLight
): UseRaterRatingContentParticipantListHook => {
  const client = useApolloClient()

  const isMounted = React.useRef(false)
  const surveyInvitationsRef = React.useRef<SurveyInvitationLight[]>([])
  const projectModulesRef = React.useRef<ProjectModule[]>([])
  const scenarioCodingItemRatingsRef = React.useRef<ScenarioCodingItemRating[]>([])
  const userRatingsRef = React.useRef<Rating[]>([])
  const raterRatingsRef = React.useRef<Rating[]>([])
  const codingDimensionsRef = React.useRef<CodingDimension[]>([])
  const ratingProjectModulesRef = React.useRef<RatingProjectModule[]>([])
  const codingCriteriaRef = React.useRef<Array<CodingCriterion | AutomatedCodingCriterion>>([])
  const codingItemsRef = React.useRef<CodingItem[]>([])
  const freetextQuestionRatingsForParticipantsRef = React.useRef<FreetextQuestionRatingsForParticipants>({})

  const [codingDimensions, setCodingDimensions] = React.useState<CodingDimension[]>([])

  const {surveyInvitations, surveyInvitationsLoading} = useSurveyInvitations(survey.id)
  const {projectModules, projectModulesLoading} = useProjectModules(survey.project.id)
  const {
    scenarioCodingItemRatings,
    scenarioCodingItemRatingsLoading,
    getScenarioCodingItemRatings
  } = useScenarioCodingItemRatingsByRatingsList()
  const {ratings, ratingsLoading} = useRatings(survey.id)
  const {codingCriteria, codingCriteriaLoading, getCodingCriteria} = useCodingCriteriaByItemsList()
  const {
    freetextQuestionRatingsForParticipants,
    freetextQuestionRatingsForParticipantsLoading,
    getFreetextQuestionRatingsForParticipants
  } = useFreetextQuestionRatingsForParticipants()

  const raterRatings = ratings.filter(({isFinalScore}) => !isFinalScore)

  const participantCount = surveyInvitationsRef.current.length

  const userRatings = React.useMemo(
    () => raterRatingsRef.current.filter(rating => rating.userAccountId === userAccountId),
    [raterRatingsRef.current, userAccountId]
  )

  const codingItems = React.useMemo(() => getCodingItemsFromCodingDimensions(codingDimensions), [codingDimensions])

  if (!isEqual(surveyInvitationsRef.current, surveyInvitations)) {
    surveyInvitationsRef.current = surveyInvitations
  }

  if (!isEqual(projectModulesRef.current, projectModules)) {
    projectModulesRef.current = projectModules
  }

  if (!isEqual(scenarioCodingItemRatingsRef.current, scenarioCodingItemRatings)) {
    scenarioCodingItemRatingsRef.current = scenarioCodingItemRatings
  }

  if (!isEqual(userRatingsRef.current, userRatings)) {
    userRatingsRef.current = userRatings
  }

  if (!isEqual(raterRatingsRef.current, raterRatings)) {
    raterRatingsRef.current = raterRatings
  }

  if (!isEqual(codingDimensionsRef.current, codingDimensions)) {
    codingDimensionsRef.current = codingDimensions
  }

  if (!isEqual(codingCriteriaRef.current, codingCriteria)) {
    codingCriteriaRef.current = codingCriteria
  }

  if (!isEqual(codingItemsRef.current, codingItems)) {
    codingItemsRef.current = codingItems
  }

  if (!isEqual(freetextQuestionRatingsForParticipantsRef.current, freetextQuestionRatingsForParticipants)) {
    freetextQuestionRatingsForParticipantsRef.current = freetextQuestionRatingsForParticipants
  }

  const isParticipantRatedByQuestionnaire = (projectModule: ProjectModule, surveyInvitation: SurveyInvitationLight) =>
    checkIsParticipantRatedByQuestionnaire({
      projectModule,
      freetextQuestionRatingsForParticipant:
        freetextQuestionRatingsForParticipantsRef.current[surveyInvitation.id] ?? [],
      ratings: userRatingsRef.current
    })

  const getRatingsByParticipants = (projectModule: ProjectModule): RatingByParticipant[] =>
    surveyInvitationsRef.current.map(surveyInvitation => {
      return {
        participantId: surveyInvitation.id,
        isNotRatable:
          projectModule.moduleType === ProjectModuleType.Scenario
            ? surveyInvitation.projectModuleProgresses.find(
                progress => progress.scenarioId === projectModule.scenarioId
              )?.status !== ProjectModuleProgressType.Completed
            : surveyInvitation.projectModuleProgresses.find(
                progress => progress.questionnaireId === projectModule.questionnaireId
              )?.status !== ProjectModuleProgressType.Completed,
        isFullyRated:
          projectModule.moduleType === ProjectModuleType.Scenario
            ? isParticipantFullyRatedByScenario({
                projectModule,
                surveyInvitation,
                codingDimensions: codingDimensionsRef.current,
                codingItems: codingItemsRef.current,
                codingCriteria: codingCriteriaRef.current,
                scenarioCodingItemRatings: scenarioCodingItemRatingsRef.current,
                ratings: userRatingsRef.current
              })
            : isParticipantRatedByQuestionnaire(projectModule, surveyInvitation)
      }
    })

  const ratingProjectModules: RatingProjectModule[] = React.useMemo(
    () =>
      sortByPosition(projectModulesRef.current).map(projectModule => {
        const ratingsByParticipants = getRatingsByParticipants(projectModule)
        const ratedParticipantCount = ratingsByParticipants.filter(entity => entity.isFullyRated || entity.isNotRatable)
          .length
        return {
          ...projectModule,
          participantCount,
          ratingsByParticipants,
          ratedParticipantCount
        }
      }),
    [
      surveyInvitationsRef.current,
      projectModulesRef.current,
      scenarioCodingItemRatingsRef.current,
      codingDimensionsRef.current,
      codingCriteriaRef.current,
      codingItemsRef.current,
      freetextQuestionRatingsForParticipantsRef.current,
      userRatingsRef.current
    ]
  )

  if (!isEqual(ratingProjectModulesRef.current, ratingProjectModules)) {
    ratingProjectModulesRef.current = ratingProjectModules
  }

  const projectModulesCount = ratingProjectModules.length

  const fullyRatedProjectModulesCount = React.useMemo(
    () =>
      sumBy(ratingProjectModules, ({participantCount, ratedParticipantCount}) =>
        participantCount === ratedParticipantCount ? 1 : 0
      ),
    [ratingProjectModules]
  )

  const fullyRatedParticipantsCount = React.useMemo(
    () =>
      sumBy(surveyInvitationsRef.current, participant => {
        const allProjectModulesRated = every(
          ({ratingsByParticipants}) =>
            find(({participantId}) => participantId === participant.id, ratingsByParticipants)
              .map(({isFullyRated, isNotRatable}) => isFullyRated || isNotRatable)
              .getOrElse(false),
          ratingProjectModules
        )
        return allProjectModulesRated ? 1 : 0
      }),
    [ratingProjectModules, surveyInvitationsRef.current]
  )

  React.useEffect(() => {
    if (userRatingsRef.current.length > 0) {
      getScenarioCodingItemRatings(userRatingsRef.current)
    }
  }, [userRatingsRef.current])

  React.useEffect(() => {
    const modules = ratingProjectModulesRef.current.filter(
      projectModule =>
        projectModule.moduleType === ProjectModuleType.Scenario &&
        projectModule.scenario !== null &&
        projectModule.scenario.codingModel !== null
    )
    Promise.all(modules.map(projectModule => getCodingDimensions(client, projectModule.scenario!.codingModel!.id)))
      .then(result => isMounted.current && setCodingDimensions(flatten(result)))
      .catch(() => isMounted.current && setCodingDimensions([]))
  }, [ratingProjectModulesRef.current])

  React.useEffect(() => {
    getCodingCriteria({items: codingItemsRef.current})
  }, [codingItemsRef.current])

  React.useEffect(() => {
    if (userRatingsRef.current.length > 0) {
      getFreetextQuestionRatingsForParticipants(
        userRatingsRef.current,
        surveyInvitationsRef.current.map(({id}) => id)
      )
    }
  }, [userRatingsRef.current, surveyInvitationsRef.current])

  React.useEffect(() => {
    isMounted.current = true
    return () => {
      isMounted.current = false
    }
  }, [])

  return {
    dataLoading:
      projectModulesLoading ||
      ratingsLoading ||
      surveyInvitationsLoading ||
      scenarioCodingItemRatingsLoading ||
      codingCriteriaLoading ||
      freetextQuestionRatingsForParticipantsLoading,
    participantCount,
    fullyRatedParticipantsCount,
    projectModulesCount,
    fullyRatedProjectModulesCount,
    surveyInvitations,
    ratingProjectModules,
    allCodingDimensions: codingDimensions
  }
}
