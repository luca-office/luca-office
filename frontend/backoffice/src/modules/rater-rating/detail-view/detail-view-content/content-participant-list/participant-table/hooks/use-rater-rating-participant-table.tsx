import * as React from "react"
import {useDispatch} from "react-redux"
import {ColumnProps} from "shared/components"
import {ProjectModuleType} from "shared/graphql/generated/globalTypes"
import {SurveyInvitationLight} from "shared/models"
import {Route as SharedRoute} from "shared/routes"
import {useLucaTranslation} from "shared/translations"
import {every, first, sortByCreatedAt, sortByPosition} from "shared/utils"
import {navigateToRouteAction} from "../../../../../../../redux/actions/navigation-action"
import {getRaterRatingParticipantTableColumns} from "../../../../../config/participant-table.config"
import {IndexedSurveyInvitation, RatingProjectModule} from "../../../../../models"

export interface UseRaterRatingParticipantTableHook {
  readonly indexedSurveyInvitations: IndexedSurveyInvitation[]
  readonly columns: ColumnProps<IndexedSurveyInvitation>[]
  readonly navigateToParticipantRatingOverview: (surveyInvitation: SurveyInvitationLight) => void
}

interface UseRaterRatingParticipantTableParams {
  readonly surveyInvitations: SurveyInvitationLight[]
  readonly ratingProjectModules: RatingProjectModule[]
}

export const useRaterRatingParticipantTable = ({
  surveyInvitations,
  ratingProjectModules
}: UseRaterRatingParticipantTableParams): UseRaterRatingParticipantTableHook => {
  const {t} = useLucaTranslation()

  const dispatch = useDispatch()

  const sortedProjectModules = React.useMemo(() => sortByPosition(ratingProjectModules), [ratingProjectModules])

  const indexedSurveyInvitations = React.useMemo(
    () =>
      // notice to use same sort as in rating/use-rating-overview
      sortByCreatedAt(surveyInvitations).map((surveyInvitation, index) => ({
        ...surveyInvitation,
        index: index + 1
      })),
    [surveyInvitations]
  )

  const fullyRatedProjectModulesCount = React.useMemo(
    () =>
      sortedProjectModules.reduce((accumulator, projectModule) => {
        const isFullyRated = every(({isFullyRated}) => isFullyRated, projectModule.ratingsByParticipants)
        return isFullyRated ? accumulator + 1 : accumulator
      }, 0),
    [sortedProjectModules]
  )

  const columns = getRaterRatingParticipantTableColumns(
    t,
    indexedSurveyInvitations,
    sortedProjectModules,
    fullyRatedProjectModulesCount
  )

  const navigateToParticipantRatingOverview = (surveyInvitation: SurveyInvitationLight) =>
    first(sortedProjectModules).forEach(projectModule =>
      dispatch(
        navigateToRouteAction(
          projectModule.moduleType === ProjectModuleType.Scenario
            ? SharedRoute.RatingOverviewSurveyModuleParticipantScenarioNormalRater
            : SharedRoute.RatingOverviewSurveyModuleParticipantQuestionnaireNormalRater,
          {
            id: surveyInvitation.survey.projectId,
            surveyId: surveyInvitation.survey.id,
            surveyInvitationId: surveyInvitation.id,
            moduleId: projectModule.id
          }
        )
      )
    )

  return {indexedSurveyInvitations, columns, navigateToParticipantRatingOverview}
}
