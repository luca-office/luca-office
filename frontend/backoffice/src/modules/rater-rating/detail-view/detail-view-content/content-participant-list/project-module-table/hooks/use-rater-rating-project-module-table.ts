import {ColumnProps} from "shared/components"
import {CodingDimension} from "shared/models"
import {useLucaTranslation} from "shared/translations"
import {getRaterRatingProjectModuleColumns} from "../../../../../config/project-module-table.config"
import {RatingProjectModule} from "../../../../../models"

export interface UseRaterRatingProjectModuleTableHook {
  readonly columns: ColumnProps<RatingProjectModule>[]
}

export interface UseRaterRatingProjectModuleTableParams {
  readonly fullyRatedParticipantCount: number
  readonly codingDimensions: CodingDimension[]
}

export const useRaterRatingProjectModuleTable = ({
  fullyRatedParticipantCount,
  codingDimensions
}: UseRaterRatingProjectModuleTableParams): UseRaterRatingProjectModuleTableHook => {
  const {t} = useLucaTranslation()

  const columns: ColumnProps<RatingProjectModule>[] = getRaterRatingProjectModuleColumns(
    t,
    codingDimensions,
    fullyRatedParticipantCount
  )

  return {columns}
}
