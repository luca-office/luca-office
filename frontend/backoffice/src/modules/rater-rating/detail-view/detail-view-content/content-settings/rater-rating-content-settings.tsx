import {css} from "@emotion/react"
import * as React from "react"
import {Paper, Text} from "shared/components"
import {AuthenticationType} from "shared/graphql/generated/globalTypes"
import {SurveyLight} from "shared/models"
import {
  CustomStyle,
  FontWeight,
  mediumBoxShadow,
  spacingMedium,
  spacingTiny,
  textEllipsis,
  TextSize
} from "shared/styles"
import {useLucaTranslation} from "shared/translations"
import {formatDateFromString} from "shared/utils"

export interface RaterRatingContentSettingsProps extends CustomStyle {
  readonly survey: SurveyLight
}

export const RaterRatingContentSettings: React.FC<RaterRatingContentSettingsProps> = ({customStyles, survey}) => {
  const {t} = useLucaTranslation()

  const conditionLabel = survey.isOpenParticipationEnabled
    ? t("projects__surveys_table_open_participation")
    : t("projects__surveys_table_closed_participation")
  const authenticationLabel =
    survey.authenticationType === AuthenticationType.OnlyAnonymous
      ? t("rater_rating_details__survey_authentication_anonymous_only")
      : survey.authenticationType === AuthenticationType.OnlyRegistered
      ? t("rater_rating_details__survey_authentication_registered_only")
      : t("rater_rating_details__survey_authentication_anonymous_allowed")

  return (
    <div css={[styles.contentWrapper, customStyles]}>
      <Text customStyles={styles.label} size={TextSize.Medium}>
        {t("rater_rating_details__settings")}
      </Text>
      <Paper customStyles={styles.settingsPaper}>
        <div css={styles.contentWrapper}>
          <Text customStyles={styles.label} size={TextSize.Medium}>
            {t("rater_rating_details__settings_conditions_of_participation")}
          </Text>
          <Text size={TextSize.Medium}>{`${conditionLabel} (${authenticationLabel})`}</Text>
        </div>
        <div css={styles.contentWrapper}>
          <Text customStyles={styles.label} size={TextSize.Medium}>
            {t("rater_rating_details__settings_survey_procedure")}
          </Text>
          <Text size={TextSize.Medium}>
            {!survey.startsAt || !survey.endsAt
              ? t("rater_rating__survey_progress_manual")
              : t("projects__surveys_details_progress_automatic", {
                  start: formatDateFromString(survey.startsAt),
                  end: formatDateFromString(survey.endsAt)
                })}
          </Text>
        </div>
      </Paper>
    </div>
  )
}

const styles = {
  contentWrapper: css({
    display: "grid",
    gridTemplateRows: "minmax(min-content, max-content) 1fr",
    gridRowGap: spacingTiny
  }),
  label: css(textEllipsis, {
    fontWeight: FontWeight.Bold
  }),
  settingsPaper: css({
    display: "grid",
    gridTemplateColumns: "repeat(2, 1fr)",
    gridColumnGap: spacingMedium,
    padding: spacingMedium,
    boxShadow: mediumBoxShadow
  })
}
