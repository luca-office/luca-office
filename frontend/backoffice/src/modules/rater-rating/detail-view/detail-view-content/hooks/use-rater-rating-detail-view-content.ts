import {useSurveyLight} from "shared/graphql/hooks/queries/survey/use-survey-light"
import {SurveyLight} from "shared/models"
import {Option} from "shared/utils"

export interface UseRaterRatingDetailViewContentHook {
  readonly dataLoading: boolean
  readonly survey: Option<SurveyLight>
}

export const useRaterRatingDetailViewContent = (surveyId: UUID): UseRaterRatingDetailViewContentHook => {
  const {survey, surveyLoading} = useSurveyLight(surveyId)

  return {dataLoading: surveyLoading, survey}
}
