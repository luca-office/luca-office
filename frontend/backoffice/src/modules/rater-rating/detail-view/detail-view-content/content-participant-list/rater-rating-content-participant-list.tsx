import {css} from "@emotion/react"
import * as React from "react"
import {LoadingIndicator, Text} from "shared/components"
import {SurveyLight} from "shared/models"
import {
  backgroundColorDarker,
  CustomStyle,
  Flex,
  FontWeight,
  spacingSmall,
  spacingTiny,
  textEllipsis,
  TextSize
} from "shared/styles"
import {useLucaTranslation} from "shared/translations"
import {TabbedCard} from "../../../../../components"
import {useRaterRatingContentParticipantList} from "./hooks/use-rater-rating-content-participant-list"
import {RaterRatingParticipantTable} from "./participant-table/rater-rating-participant-table"
import {RaterRatingProjectModuleTable} from "./project-module-table/rater-rating-project-module-table"

export interface RaterRatingContentParticipantListProps extends CustomStyle {
  readonly userAccountId: UUID
  readonly survey: SurveyLight
}

export const RaterRatingContentParticipantList: React.FC<RaterRatingContentParticipantListProps> = ({
  customStyles,
  userAccountId,
  survey
}) => {
  const {t} = useLucaTranslation()
  const {
    dataLoading,
    participantCount,
    fullyRatedParticipantsCount,
    projectModulesCount,
    fullyRatedProjectModulesCount,
    surveyInvitations,
    ratingProjectModules,
    allCodingDimensions
  } = useRaterRatingContentParticipantList(userAccountId, survey)

  const loadingIndicator = (
    <div css={styles.loadingIndicatorWrapper}>
      <LoadingIndicator />
    </div>
  )

  return (
    <div css={[styles.content, customStyles]}>
      <Text customStyles={styles.contentLabel} size={TextSize.Medium}>
        {t("rater_rating_details__rating_list")}
      </Text>
      <TabbedCard
        defaultActiveIndex={0}
        tabs={[
          {
            label: `${t("dashboard__progress_chart_legend_open")} (${fullyRatedParticipantsCount}/${participantCount})`,
            content: dataLoading ? (
              loadingIndicator
            ) : (
              <RaterRatingParticipantTable
                surveyInvitations={surveyInvitations}
                ratingProjectModules={ratingProjectModules}
              />
            )
          },
          {
            label: `${t(
              "rater_rating_details__project_modules"
            )} (${fullyRatedProjectModulesCount}/${projectModulesCount})`,
            content: dataLoading ? (
              loadingIndicator
            ) : (
              <RaterRatingProjectModuleTable
                fullyRatedParticipantCount={fullyRatedParticipantsCount}
                codingDimensions={allCodingDimensions}
                ratingProjectModules={ratingProjectModules}
              />
            )
          }
        ]}
      />
    </div>
  )
}

const styles = {
  content: css({
    display: "grid",
    gridTemplateRows: "minmax(min-content, max-content) 1fr",
    gridRowGap: spacingTiny
  }),
  contentLabel: css(textEllipsis, {
    fontWeight: FontWeight.Bold
  }),
  loadingIndicatorWrapper: css(Flex.column, {
    justifyContent: "center",
    alignItems: "center",
    padding: spacingSmall,
    backgroundColor: backgroundColorDarker
  })
}
