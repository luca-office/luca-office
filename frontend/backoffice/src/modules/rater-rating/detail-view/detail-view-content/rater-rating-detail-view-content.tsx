import {css} from "@emotion/react"
import * as React from "react"
import {LoadingIndicator} from "shared/components"
import {
  backgroundColorBright,
  borderRadius,
  spacing,
  spacingHuge,
  spacingLarge,
  spacingMedium,
  spacingSmall
} from "shared/styles"
import {RaterRatingContentInfo} from "./content-info/rater-rating-content-info"
import {RaterRatingContentParticipantList} from "./content-participant-list/rater-rating-content-participant-list"
import {RaterRatingContentSettings} from "./content-settings/rater-rating-content-settings"
import {useRaterRatingDetailViewContent} from "./hooks/use-rater-rating-detail-view-content"

export interface RaterRatingDetailViewContentProps {
  readonly userAccountId: UUID
  readonly surveyId: UUID
}

export const RaterRatingDetailViewContent: React.FC<RaterRatingDetailViewContentProps> = ({
  userAccountId,
  surveyId
}) => {
  const {dataLoading, survey: surveyOption} = useRaterRatingDetailViewContent(surveyId)

  return (
    <div css={styles.container}>
      <div css={styles.content(dataLoading)}>
        {dataLoading ? (
          <LoadingIndicator />
        ) : (
          surveyOption
            .map(survey => (
              <React.Fragment>
                <RaterRatingContentInfo survey={survey} />
                <RaterRatingContentSettings customStyles={styles.settings} survey={survey} />
                <RaterRatingContentParticipantList
                  customStyles={styles.participantList}
                  userAccountId={userAccountId}
                  survey={survey}
                />
              </React.Fragment>
            ))
            .orNull()
        )}
      </div>
    </div>
  )
}

const styles = {
  container: css({
    padding: spacing(spacingMedium, spacingHuge),
    overflow: "auto"
  }),
  content: (loading: boolean) =>
    css({
      ...(loading
        ? {display: "flex", justifyContent: "center", alignItems: "center"}
        : {display: "grid", gridTemplateRows: "repeat(2, minmax(min-content, max-content)) 1fr"}),
      padding: spacing(spacingLarge, spacingMedium, spacingMedium, spacingMedium),
      borderRadius: borderRadius,
      backgroundColor: backgroundColorBright
    }),
  settings: css({
    marginTop: spacingHuge + spacingSmall
  }),
  participantList: css({
    marginTop: spacingLarge
  })
}
