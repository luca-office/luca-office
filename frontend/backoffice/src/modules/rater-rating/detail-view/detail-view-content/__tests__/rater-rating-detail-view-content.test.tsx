import {MockedProvider} from "@apollo/client/testing"
import {shallow} from "enzyme"
import * as React from "react"
import {create} from "react-test-renderer"
import {LoadingIndicator} from "shared/components"
import {surveyIdMock, surveyInvitationsMock, surveyLightMock, userAccountMock} from "shared/graphql/__mocks__"
import {surveyInvitationsQuery} from "shared/graphql/queries"
import {Option} from "shared/utils"
import {RaterRatingContentInfo} from "../content-info/rater-rating-content-info"
import {RaterRatingContentParticipantList} from "../content-participant-list/rater-rating-content-participant-list"
import {RaterRatingContentSettings} from "../content-settings/rater-rating-content-settings"
import * as useRaterRatingDetailviewContentHook from "../hooks/use-rater-rating-detail-view-content"
import {UseRaterRatingDetailViewContentHook} from "../hooks/use-rater-rating-detail-view-content"
import {RaterRatingDetailViewContent, RaterRatingDetailViewContentProps} from "../rater-rating-detail-view-content"

const defaultProps: RaterRatingDetailViewContentProps = {
  userAccountId: userAccountMock.id,
  surveyId: surveyIdMock
}

const stateHookValuesDefault: UseRaterRatingDetailViewContentHook = {
  dataLoading: false,
  survey: Option.of(surveyLightMock)
}

const stateSpy = jest.spyOn(useRaterRatingDetailviewContentHook, "useRaterRatingDetailViewContent")

const getComponent = (props?: Partial<RaterRatingDetailViewContentProps>) => (
  <MockedProvider
    mocks={[
      {
        request: {
          query: surveyInvitationsQuery,
          variables: {surveyId: surveyIdMock}
        },
        result: {
          data: {
            surveyInvitations: surveyInvitationsMock
          }
        }
      }
    ]}>
    <RaterRatingDetailViewContent {...{...defaultProps, ...props}} />
  </MockedProvider>
)

describe("rater-rating-detail-view-content", () => {
  it("renders correctly", () => {
    stateSpy.mockReturnValue(stateHookValuesDefault)
    const component = getComponent()
    const tree = create(component).toJSON()
    expect(tree).toMatchSnapshot()
  })
  it("renders correctly (loading)", () => {
    stateSpy.mockReturnValue({...stateHookValuesDefault, dataLoading: true})
    const component = getComponent()
    const tree = create(component).toJSON()
    expect(tree).toMatchSnapshot()
  })
  it("has correct structure", () => {
    stateSpy.mockReturnValue(stateHookValuesDefault)
    const component = getComponent()
    const tree = shallow(component).childAt(0).dive()

    expect(tree.find(LoadingIndicator)).toHaveLength(0)
    expect(tree.find(RaterRatingContentInfo)).toHaveLength(1)
    expect(tree.find(RaterRatingContentSettings)).toHaveLength(1)
    expect(tree.find(RaterRatingContentParticipantList)).toHaveLength(1)
  })
  it("has correct structure (loading)", () => {
    stateSpy.mockReturnValue({...stateHookValuesDefault, dataLoading: true})
    const component = getComponent()
    const tree = shallow(component).childAt(0).dive()

    expect(tree.find(LoadingIndicator)).toHaveLength(1)
    expect(tree.find(RaterRatingContentInfo)).toHaveLength(0)
    expect(tree.find(RaterRatingContentSettings)).toHaveLength(0)
    expect(tree.find(RaterRatingContentParticipantList)).toHaveLength(0)
  })
})
