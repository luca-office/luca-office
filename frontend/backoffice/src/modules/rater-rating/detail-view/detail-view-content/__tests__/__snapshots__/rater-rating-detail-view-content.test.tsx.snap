// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`rater-rating-detail-view-content renders correctly (loading) 1`] = `
<div
  className="css-em4hye-RaterRatingDetailViewContent"
>
  <div
    className="css-1ij374s-RaterRatingDetailViewContent"
  >
    <div
      className="css-1tzcb72-Icon"
      dangerouslySetInnerHTML={
        {
          "__html": {},
        }
      }
    />
  </div>
</div>
`;

exports[`rater-rating-detail-view-content renders correctly 1`] = `
<div
  className="css-em4hye-RaterRatingDetailViewContent"
>
  <div
    className="css-1x3womu-RaterRatingDetailViewContent"
  >
    <div
      className="css-mns6bu-RaterRatingContentInfo"
    >
      <div
        className="css-118ld05-RaterRatingContentInfo"
      >
        <div
          className="css-1f1k13x-Text"
        >
          rater_rating_details__title
        </div>
        <div
          className="css-wmw0az-RaterRatingContentInfo"
        >
          <div
            className="css-tyag6n-Heading"
          >
            my test scenario
          </div>
          <div
            className="css-1xkyok3-Heading"
          >
            dashboard__title_time_ended
          </div>
        </div>
      </div>
      <div
        className="css-118ld05-RaterRatingContentInfo"
      >
        <div
          className="css-1f1k13x-Text"
        >
          rater_rating_details__description
        </div>
        <div
          className="css-1rh0ks6-Text"
        >
          description
        </div>
      </div>
    </div>
    <div
      className="css-1hyj09g-RaterRatingContentSettings"
    >
      <div
        className="css-1f1k13x-Text"
      >
        rater_rating_details__settings
      </div>
      <div>
        <div
          className="css-1t5zhaa-Paper"
        >
          <div
            className="css-hdaeu0-RaterRatingContentSettings"
          >
            <div
              className="css-1f1k13x-Text"
            >
              rater_rating_details__settings_conditions_of_participation
            </div>
            <div
              className="css-17zbm7y-Text"
            >
              projects__surveys_table_closed_participation (rater_rating_details__survey_authentication_registered_only)
            </div>
          </div>
          <div
            className="css-hdaeu0-RaterRatingContentSettings"
          >
            <div
              className="css-1f1k13x-Text"
            >
              rater_rating_details__settings_survey_procedure
            </div>
            <div
              className="css-17zbm7y-Text"
            >
              projects__surveys_details_progress_automatic
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      className="css-175pgxb-RaterRatingContentParticipantList"
    >
      <div
        className="css-1f1k13x-Text"
      >
        rater_rating_details__rating_list
      </div>
      <div
        className="css-63bajj-Card"
      >
        <div
          className="card-header css-y8kuyc-CardHeader"
        >
          <div
            className="css-1fcutbn-TabbedCard"
          >
            <div
              className="card-tab css-dqywdv-Text"
              onClick={[Function]}
            >
              dashboard__progress_chart_legend_open (0/0)
            </div>
            <div
              className="card-tab css-182t43y-Text"
              onClick={[Function]}
            >
              rater_rating_details__project_modules (0/0)
            </div>
          </div>
        </div>
        <div
          className="card-content css-1vyr92j-CardContent"
        >
          <div
            className="card-tab-content css-o4mi6l-TabbedCard"
          >
            <div
              className="css-21ikin-RaterRatingContentParticipantList"
            >
              <div
                className="css-1tzcb72-Icon"
                dangerouslySetInnerHTML={
                  {
                    "__html": {},
                  }
                }
              />
            </div>
          </div>
          <div
            className="card-tab-content css-16oetws-TabbedCard"
          >
            <div
              className="css-21ikin-RaterRatingContentParticipantList"
            >
              <div
                className="css-1tzcb72-Icon"
                dangerouslySetInnerHTML={
                  {
                    "__html": {},
                  }
                }
              />
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
